src/output/plugins/visualization/SoundAnalysis.hxx

```diff
@@ -0,0 +1,295 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Copyright The Music Player Daemon Project
+
+#ifndef SOUND_ANALYSIS_HXX_INCLUDED
+#define SOUND_ANALYSIS_HXX_INCLUDED 1
+
+#include "SoundInfoCache.hxx"
+#include "LowLevelProtocol.hxx"
+
+#include <fftw3.h>
+
+#include <algorithm>
+#include <cstddef>
+#include <memory>
+
+#include <math.h>
+
+struct ConfigBlock;
+
+namespace Visualization {
+
+/**
+ * \brief Convenience class for expressing sound analysis parameters exclusive
+ * of the audio format
+ *
+ *
+ * There are any number of parameters governing our analysis of PCM data. Other
+ * than the `AudioFormat`, they are read from configuration at startup and
+ * constant. Rather than force callers to write methods taking many parameters,
+ * this class colects them all in one place, and enforces some constraints on
+ * their values.
+ *
+ *
+ */
+
+class SoundAnalysisParameters
+{
+	/* The number of samples used for each analysis; this must be greater than
+	 * zero and needn't be large (say, less than 1024). Configuration value
+	 * "num_samples" */
+	size_t num_samples;
+	/* Data lower than this frequency (in the frequency domain) shall be
+	 * discarded; must be greater than or equal to zero, and less than
+	 * hi_cutoff. A typical value would be 200 (the lower range of human
+	 * perception). Units: Hz. Configuration value "lo_cutoff" */
+	float lo_cutoff;
+	/* Data greater than this frequency (in the frequency domain) shall be
+	 * discarded; must be greater than or equal to zero, and greater than
+	 * lo_cutoff. A typical value would be 10000-12000 (empirically, there's not a
+	 * lot of activity above 10000 in song data). Units Hz. Configuration value
+	 * "hi_cutoff" */
+	float hi_cutoff;
+
+	static constexpr size_t DEF_NUM_SAMPLES = 513;
+	static constexpr size_t DEF_LO_CUTOFF = 200;
+	static constexpr size_t DEF_HI_CUTOFF = 10000;
```

Comment:
What does `DEF_` mean and why is this a `size_t` when the non-static fields with similar names are `float`?

Reply:
Short for "default"-- re-named. They're integral because `ConfigBlock` offers no accessors for floating-point values.